Migration of Advanced Parameters > Team > Profiles page.

rokaszygmantas opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates the profiles list page (Advanced Parameters > Team > Profiles).
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10505
| How to test?  | Compare to the legacy controller. This PR migrates only the listing page of Profiles. URL: `admin-dev/index.php/configure/advanced/profiles/`. **You have to rebuild the assets before testing.** Superadmin specific rules will be done in another PR (see https://github.com/PrestaShop/PrestaShop/issues/10505)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

- [x] Create grid
- [x] Implement grid actions
- [x] Specific logic on grid bulk/row actions

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10783)
<!-- Reviewable:end -->

rokaszygmantas committed:
Add profiles grid, without actions

rokaszygmantas committed:
Fix profiles grid javascript actions

rokaszygmantas committed:
Fix pagination in profiles grid

mickaelandrieu commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -0,0 +1,161 @@
+        $definitionFactory = $definitionFactory->getDefinition();
+
+        $gridFilterFormFactory = $this->get('prestashop.core.grid.filter.form_factory');
+        $searchParametersForm = $gridFilterFormFactory->create($definitionFactory);
```
should be createForm($definitionFactory), I've forgotten to update it :/

sarjon replied:
:+1: for `createForm` and not `getForm`. :p

PierreRambaud replied:
Why redundant `Form`, a form factory can only create form... `create` is the right name

sarjon replied:
@PierreRambaud where were you when we had discussion with @mickaelandrieu about that for other services. -_- i needed your backup. :p

PierreRambaud replied:
Didn't see the discussion... Too many things to do :'(

sarjon replied:
in short, it was about `GridDefinitionFactoryInterface::create()` => `GridDefinitionFactoryInterface::getDefinition()`

PierreRambaud replied:
Always call me when you want someone to do mess with naming :smile:

sarjon replied:
we may still fix that untill 1.7.5 release if you get @mickaelandrieu on your side. :smile:

PierreRambaud replied:
@matks wdyt too?

matks replied:
I'm sorry, I'm going to disappoint all of you ^^ : I think both names are right. One is more redundant, but both names provide the required information ("what is this function doing ?") very accurately so I do not care.

mickaelandrieu replied:
I think it's too late to change the name, but if we can I stand for `getForm()` for consistency with all others functions in the Grid component.

tomas862 commented on src/Core/Grid/Definition/Factory/ProfilesGridDefinitionFactory.php:
```diff
@@ -0,0 +1,207 @@
+     */
+    protected function getId()
+    {
+        return 'profiles';
```
All new grid ids starts with  upper-case 

```suggestion
        return 'Profiles';
```

mickaelandrieu replied:
is it a convention? If yes, it needs to be documented.

tomas862 commented on src/Core/Grid/Query/ProfilesQueryBuilder.php:
```diff
@@ -0,0 +1,125 @@
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicatorInterface $searchCriteriaApplicator,
+        $languageId
```
Maybe a better naming of this id would be $contextLanguageId ? :thinking: 
```suggestion
        $contextLanguageId
```

tomas862 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -0,0 +1,161 @@
+     * Shows profile edit form.
+     *
+     * @AdminSecurity(
+     *     "is_granted('update', request.get('_legacy_controller'))",
```
Provide redirectRoute parameter

```suggestion
     *     "is_granted('update', request.get('_legacy_controller'))",
     *     redirectRoute="admin_profiles",
```

tomas862 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -0,0 +1,161 @@
+     * Delete a profile.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['delete'], request.get('_legacy_controller')~'_')",
```
```suggestion
     *     "is_granted('delete', request.get('_legacy_controller')~'_')",
```

tomas862 commented on src/PrestaShopBundle/Resources/config/routing/admin/configure/advanced_parameters/profiles.yml:
```diff
@@ -0,0 +1,38 @@
+admin_profiles:
```
Check naming convention :+1: 
```suggestion
admin_profiles_index:
```

tomas862 commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -0,0 +1,161 @@
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @Template("@PrestaShop/Admin/Configure/AdvancedParameters/Profiles/profiles.html.twig")
```
Check naming convention :+1: 

```suggestion
     * @Template("@PrestaShop/Admin/Configure/AdvancedParameters/Profiles/index.html.twig")
```

rokaszygmantas committed:
Fix action/route naming

rokaszygmantas committed:
Move accessibility checker to core namespace

rokaszygmantas committed:
Split grid services definitions to separate files

rokaszygmantas committed:
Implement single profile deletion

rokaszygmantas committed:
Implement bulk profile deletion

rokaszygmantas committed:
Fix cs

rokaszygmantas commented:
Hi @matks,

This listing page migrated, except there is one thing that is different compared to legacy controller, and it can be seen in these screenshots:
Legacy: https://prnt.sc/lxvy86
Migrated: https://prnt.sc/lxvy2v
The bulk actions are rendered for all rows and the accessibility checking is not implemented for bulk actions in the grid component yet. I've talked with @sarjon about this and we came to a conclusion that it would be better to open another PR with this improvement to grid component, because it's a global change needed for many grid implementations. Wdyt?

Btw the backend accessibility checking is implemented, so you cannot delete the super admin profile, it's just that the checkbox is displayed visually.

matks commented:
@rokaszygmantas Great 👍 

Yes it's nice to split deliverables into several PRs. I added this item to [the issue todlist](https://github.com/PrestaShop/PrestaShop/issues/10505#issue-361803488) so we dont forget.

Now I've got to review 😄

matks commented on src/Core/Grid/Query/ProfilesQueryBuilder.php:
```diff
@@ -0,0 +1,125 @@
+            ->setParameter('language', $this->languageId)
+        ;
+
+        foreach ($filters as $name => $value) {
```
Please restrict the possible `$filters` value. We know this reduces the reusability of this class, but security is more important
```
    $allowedFilters = [ ... ];

     foreach ($filters as $name => $value) {
            if (!in_array($filterName, $allowedFilters, true)) {
                continue;
            }
```

matks commented on src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfilesController.php:
```diff
@@ -0,0 +1,204 @@
+    /**
+     * Show profiles listing page.
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
```
We have an ongoing debate about whether we should add `create`, `update` and `delete` rights here
https://github.com/PrestaShop/PrestaShop/issues/12109

matks replied:
We decided that `read` right should be the only one giving access to listing page. So this code is valid 😄 👍

matks reviewed (commented):
Almost nothing more to add than Tomas and Mickael 😄

matks commented:
@rokaszygmantas (finally) reviewed ! sorry for the wait ...

rokaszygmantas committed:
Restore accessibility checker on row action

rokaszygmantas commented:
@matks rebased, everything is done from my side on this PR. :slightly_smiling_face:

matks commented:
@rokaszygmantas I'm sorry I dont see the fix for https://github.com/PrestaShop/PrestaShop/pull/10783/files#r247077480 🤔

rokaszygmantas commented:
@matks sorry, must have missed it! I'll fix it and review other comments once more :)

rokaszygmantas committed:
Check if filter name is allowed before using it in filtering query

rokaszygmantas committed:
Minor fixes

rokaszygmantas commented:
@matks fixed it :)

marionf commented:
@TristanLDD 

Is it ok for you ?

![capture d ecran_942](https://user-images.githubusercontent.com/13449658/51267490-29f65180-19be-11e9-90c4-49ef30fc7fbc.png)

TristanLDD commented:
Hello,

2 little fixes on this page: 

- Add placeholders in search fields 

-Reduce the ID search field to 150px width 

Thank you !

rokaszygmantas committed:
Make class final

rokaszygmantas committed:
Add search placeholders

rokaszygmantas committed:
Add a new common form type for grid search by ID

rokaszygmantas committed:
Fix typo

rokaszygmantas commented:
Hi @marionf, I've fixed the PR according to feedback from @TristanLDD (added placeholders and fixed ID input width). :slightly_smiling_face: 

@matks - I've added a new form type for grid search by ID (`\PrestaShopBundle\Form\Admin\Type\SearchByIdType`), which will have reduced width automatically, so it can be reused across other grids for consistency, wdyt?

rokaszygmantas committed:
Fix rebasing mistake

matks commented:
@TristanLDD Screenshot after Rokas changes:
![capture d ecran 2019-01-24 a 17 56 15](https://user-images.githubusercontent.com/3830050/51694562-669afc00-2001-11e9-9e66-84549521013c.png)

matks commented:
@rokaszygmantas Mmmm ... isn't it a bit overkill to add a new FilterType only for CSS display ? 😅 

Maybe we can make it rather an option of current `TextType`, something like
```
->setTypeOptions([
                    'size' => 'small' // we could choose either 'small', 'big' or 'full', full is to use the whole space
                ])
```

rokaszygmantas commented:
oh well, I liked the idea of having a separate type :smile: but I can change it to a size option for `TextType`, not a big deal :slightly_smiling_face:

matks commented:
@rokaszygmantas If the new Type defines a different backend behavior, why not. But here its only purpose is to be used as a marker for specific CSS. In theory, the Grid Component is not display-related (it's rather: we have a Grid component and we built on top of it a templating system using Twig).

Moreover, this "size" option will become useful if tomorrow we get to migrate a grid with a "ISO Code" column for example (so max characters length is 3) and Design Team asks us to make it small as well. You can't use a SearchById Type for a SearchByIsoCode field, right 😉 ?

rokaszygmantas commented:
@matks nice explanation - makes sense!

TristanLDD commented:
@matks Thank you, design is Ok for me

rokaszygmantas committed:
Remove SearchById form type

rokaszygmantas committed:
Introduce size option for text inputs

rokaszygmantas committed:
Fix cs

rokaszygmantas commented:
Hi @matks, 

I've removed the `SearchById` form type now and introduced new `size` option for `TextType` types. 

I didn't implement other size options besides `small`, they can be easily added when needed. :slightly_smiling_face:

matks commented:
I think we dont need to bother QA to review the UX feedback 😄

matks commented:
Thank you @rokaszygmantas !
